### ISSUE TITLE
No signed integer overflow warning reported in the call to constructor

### ISSUE THREAD

**Issue by Fedr** [llvm/llvm-project#58944]
In the following program there are two similar signed integer overflows:
```
#include <cstdint>

void func(uint64_t) {}

struct A { 
    A(uint64_t){} 
};

int main() {
    func(1024 * 1024 * 1024 * 1024 * 1024ull); // warning, ok
    A a(1024 * 1024 * 1024 * 1024 * 1024ull); // no warning in Clang!
}
```
The one in calling `func` is detected by Clang.
But the other one in calling `A::A` does not, even with `-Wall -Wextra` flags. At the same time both GCC and MSVC reports the warning here as well. Online demo: https://gcc.godbolt.org/z/6vGT33q1W

Related discussion: https://stackoverflow.com/q/73962573/7325599

**Comment by shafik** [comment]
So we are hitting `CheckedIntArithmetic(...)` in `ExprConstant.cpp` in both cases but for the second case it looks like `CheckingForUndefinedBehavior` is `false` and therefore no diagnostic.

**Comment by shafik** [comment]
In the first case we end up going through `Expr::EvaluateForOverflow(const ASTContext &Ctx)` which explicitly set it:

```cpp
Info.CheckingForUndefinedBehavior = true;
```

So it looks like `Sema::CheckForIntOverflow (Expr *E)` is catching the `CallExpr` case which covers `func(1024 * 1024 * 1024 * 1024 * 1024ull)` but does not handle the `CXXConstructExpr` case which would handle the other case.

I did a quick fix and added:

```cpp
else if (auto Construct = dyn_cast<CXXConstructExpr>(E))
      Exprs.append(Construct->arg_begin(), Construct->arg_end());
``` 

and that seems to fix the problem. If this passes a `check-clang` I will try and post a patch.

**Comment by shafik** [comment]
PR: https://reviews.llvm.org/D137897
